patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs

```diff
@@ -125,62 +126,157 @@ internal static class WorldIO
 
 		internal static List<TagCompound> SaveNPCs() {
 			var list = new List<TagCompound>();
-			for (int k = 0; k < Main.npc.Length; k++) {
-				NPC npc = Main.npc[k];
-				if (npc.active && NPCLoader.IsModNPC(npc)) {
-					if (npc.townNPC) {
-						TagCompound tag = new TagCompound {
-							["mod"] = npc.ModNPC.Mod.Name,
-							["name"] = npc.ModNPC.Name,
-							["displayName"] = npc.GivenName,
-							["x"] = npc.position.X,
-							["y"] = npc.position.Y,
-							["homeless"] = npc.homeless,
-							["homeTileX"] = npc.homeTileX,
-							["homeTileY"] = npc.homeTileY
-						};
-						list.Add(tag);
+			var data = new TagCompound();
+			
+			for (int index = 0; index < Main.maxNPCs; index++) {
+				NPC npc = Main.npc[index];
+
+				if (!npc.active || (!npc.townNPC || npc.type == NPCID.TravellingMerchant) && !NPCLoader.SavesAndLoads(npc)) {
+					continue;
+				}
+
+				var globalData = new List<TagCompound>();
+
+				foreach (GlobalNPC globalNPC in npc.globalNPCs.Select(instancedGlobalNPC => instancedGlobalNPC.Instance)) {
+					if (globalNPC is UnloadedGlobalNPC unloadedGlobalNPC) {
+						globalData.AddRange(unloadedGlobalNPC.data);
+						continue;
 					}
-					else if (NPCID.Sets.SavesAndLoads[npc.type]) {
-						TagCompound tag = new TagCompound {
-							["mod"] = npc.ModNPC.Mod.Name,
-							["name"] = npc.ModNPC.Name,
-							["x"] = npc.position.X,
-							["y"] = npc.position.Y
-						};
-						list.Add(tag);
+
+					globalNPC.SaveData(npc, data);
+
+					if (data.Count != 0) {
+						globalData.Add(new TagCompound {
+							["mod"] = globalNPC.Mod.Name,
+							["name"] = globalNPC.Name,
+							["data"] = data
+						});
+
+						data = new TagCompound();
 					}
 				}
+
+				TagCompound tag;
+
+				if (NPCLoader.IsModNPC(npc)) {
```

Comment:
should probably check `modNPC.NeedsSaving` here, or you'll confuse people about why `SaveData` gets called even when they said `NeedsSaving` is false, just because there's a global or some other reason the NPC is being saved

Reply:
The whole logic behind the NPC save logic is that NPCs can be saved even if the modNPC saving is set to false, and town NPC are a special case where you always save (mainly because it's the previous behavior and everyone wants their town NPC to be saved)